File: web/src/api/tasks.ts
// Typed client for GET /projects/:id/tasks (#125), served for real by
// cmd/api / cmd/local now — previously this only worked against
// web/mock-server.mjs (see git history). The backend returns tasks flat,
// linked by parent_id, exactly as internal/domain.Task shapes them; this
// module is what turns that into the nested `subtasks` tree the rest of
// the app (ProjectDetailPage, WhatsNextCard) renders.
import { request } from '@/src/api/projects'

// Task mirrors internal/domain.Task field-for-field, plus a client-side
// `subtasks` array built by buildTaskTree — the wire response has no such
// field.
export interface Task {
  id: string
  project_id: string
  parent_id: string
  title: string
  description: string
  status: string
  order: number
  acceptance_criteria: string[]
  subtasks: Task[]
}

// FlatTask is the wire shape: internal/domain.Task as-is, no subtasks.
// Exported so other clients returning the same shape (e.g.
// web/src/api/skills.ts's accept-changeset response) can reuse it instead
// of redefining it.
export type FlatTask = Omit<Task, 'subtasks'>

// buildTaskTree groups a flat, parent_id-linked list into a tree: tasks
// with no parent_id (or one that isn't present in the list) are roots.
// Each level is sorted by order, matching how the backend orders subtasks
// within their parent.
export function buildTaskTree(flat: FlatTask[]): Task[] {
  const byParent = new Map<string, FlatTask[]>()
  const ids = new Set(flat.map((task) => task.id))
  for (const task of flat) {
    const parentKey = task.parent_id && ids.has(task.parent_id) ? task.parent_id : ''
    const siblings = byParent.get(parentKey) ?? []
    siblings.push(task)
    byParent.set(parentKey, siblings)
  }

  function attach(parentKey: string): Task[] {
    const siblings = byParent.get(parentKey) ?? []
    return siblings
      .map((task) => ({ ...task, subtasks: attach(task.id) }))
      .sort((a, b) => a.order - b.order)
  }

  return attach('')
}

export function flattenTasks(tasks: Task[]): Task[] {
  return tasks.flatMap((task) => [task, ...flattenTasks(task.subtasks)])
}

export async function listTasks(projectId: string): Promise<Task[]> {
  const body = await request<FlatTask[]>(`/api/projects/${encodeURIComponent(projectId)}/tasks`)
  return buildTaskTree(body)
}

// CreateTaskRequest mirrors docs/openapi.yaml's CreateTaskRequest — a
// single task added by hand (#175), outside decompose_task's
// propose/accept flow. parent_id nests it under an existing task.
export interface CreateTaskRequest {
  title: string
  description?: string
  acceptance_criteria?: string[]
  parent_id?: string
}

export function createTask(projectId: string, input: CreateTaskRequest): Promise<FlatTask> {
  return request<FlatTask>(`/api/projects/${encodeURIComponent(projectId)}/tasks`, {
    method: 'POST',
    body: JSON.stringify(input),
  })
}

File: internal/api/tasks.go
package api

import (
	"encoding/json"
	"errors"
	"net/http"
	"strings"

	"github.com/esvarez/lucas-assist/internal/auth"
	"github.com/esvarez/lucas-assist/internal/domain"
	"github.com/esvarez/lucas-assist/internal/store"
)

// listTasksHandler backs GET /projects/{id}/tasks (#125). It returns the
// project's tasks flat, exactly as domain.Task and store.Repository.
// ListTasks already shape them (ParentID-linked, not nested) — the same
// convention every other route in this package follows for domain.Project
// and domain.Changeset: no response-shape transformation here. Building a
// tree out of the flat list, if a caller wants one, is presentation logic
// that belongs on the client (see web/src/api/tasks.ts).
//
// The project is fetched first, the same ownership check
// acceptChangesetHandler does, so a nonexistent or another user's project
// ID 404s instead of silently returning an empty task list —
// store.Repository.ListTasks itself never returns ErrNotFound, since it
// only queries a key prefix and has no way to distinguish "no tasks yet"
// from "no such project."
func listTasksHandler(repo ProjectRepository) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		userID, _ := auth.UserIDFromContext(r.Context())
		projectID := r.PathValue("id")

		if _, err := repo.GetProject(r.Context(), userID, projectID); err != nil {
			if errors.Is(err, store.ErrNotFound) {
				writeJSON(w, http.StatusNotFound, map[string]string{"error": err.Error()})
				return
			}
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}

		tasks, err := repo.ListTasks(r.Context(), userID, projectID)
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}

		writeJSON(w, http.StatusOK, tasks)
	}
}

// createTaskRequest is a single task added by hand (#175), outside the
// decompose_task propose/accept flow. ParentID, when set, makes it a
// subtask of an existing task in the same project.
type createTaskRequest struct {
	Title              string   `json:"title" validate:"required"`
	Description        string   `json:"description"`
	AcceptanceCriteria []string `json:"acceptance_criteria"`
	ParentID           string   `json:"parent_id"`
}

// createTaskHandler backs POST /projects/{id}/tasks (#175). It's a plain
// CRUD write of content the user typed themselves — no model output is
// involved, so it goes straight to store.Repository.CreateTask rather
// than through a changeset (AGENTS.MD rule 1 only covers LLM proposals).
//
// Same ownership check as listTasksHandler: the project is fetched first
// so a nonexistent or another user's project ID 404s. A parent_id is
// looked up the same way, scoped to this user and project, so a subtask
// can never be attached under a task it couldn't otherwise see.
func createTaskHandler(repo ProjectRepository) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var req createTaskRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid request body: " + err.Error()})
			return
		}

		// Trimmed before validating so a whitespace-only title fails
		// "required" instead of creating a blank-looking task.
		req.Title = strings.TrimSpace(req.Title)
		req.Description = strings.TrimSpace(req.Description)
		req.ParentID = strings.TrimSpace(req.ParentID)
		criteria := make([]string, 0, len(req.AcceptanceCriteria))
		for _, c := range req.AcceptanceCriteria {
			if c = strings.TrimSpace(c); c != "" {
				criteria = append(criteria, c)
			}
		}

		if !validateStruct(w, req) {
			return
		}

		userID, _ := auth.UserIDFromContext(r.Context())
		projectID := r.PathValue("id")

		if _, err := repo.GetProject(r.Context(), userID, projectID); err != nil {
			if errors.Is(err, store.ErrNotFound) {
				writeJSON(w, http.StatusNotFound, map[string]string{"error": err.Error()})
				return
			}
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}

		if req.ParentID != "" {
			if _, err := repo.GetTask(r.Context(), userID, projectID, req.ParentID); err != nil {
				if errors.Is(err, store.ErrNotFound) {
					writeJSON(w, http.StatusBadRequest, validationErrorResponse{
						Error:  "validation failed",
						Fields: map[string]string{"parent_id": "parent task not found in this project"},
					})
					return
				}
				writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
				return
			}
		}

		// Appended after its siblings — the list is sorted by order within
		// each parent (web/src/api/tasks.ts buildTaskTree), so a hand-added
		// task lands at the end rather than colliding with an existing one.
		existing, err := repo.ListTasks(r.Context(), userID, projectID)
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}
		order := 0
		for _, t := range existing {
			if t.ParentID == req.ParentID && t.Order >= order {
				order = t.Order + 1
			}
		}

		created, err := repo.CreateTask(r.Context(), userID, domain.Task{
			ProjectID:   projectID,
			ParentID:    req.ParentID,
			Title:       req.Title,
			Description: req.Description,
			// "todo" for the same reason AcceptChangeset uses it: it's the
			// unstarted status the rest of the app treats as eligible.
			Status:             "todo",
			Order:              order,
			AcceptanceCriteria: criteria,
		})
		if err != nil {
			if errors.Is(err, store.ErrDuplicateID) {
				writeJSON(w, http.StatusConflict, map[string]string{"error": err.Error()})
				return
			}
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}

		writeJSON(w, http.StatusCreated, created)
	}
}
